### DISCUSSION TITLE
Memory usage

### DISCUSSION THREAD

**Discussion by wtholliday** [elalish/manifold#960] [General]
Starting a thread here about my memory usage explorations. Currently my app gets killed often by iOS for using too much memory during a boolean. Here's what a graph looks like for a successful boolean:

<img width="482" alt="image" src="https://github.com/user-attachments/assets/7c2c26cb-09e5-442b-a1bf-2e12b11543c1">

So high-water-mark memory usage is ~3x steady state during the boolean.

Biggest single allocation (1.18 gb):

<img width="397" alt="image" src="https://github.com/user-attachments/assets/6211ca4a-7ce0-41da-adc1-dec6e5c9bf74">

**Comment by wtholliday** [comment]
So `SparseIndices` is a (sorted) array of pairs of indices: `{p1: q1, p2: q2, ...}`

Are the p's unique? I'm guessing not, since it uses `stable_sort`.  If this were my own code (i.e. free of library-user constraints) I'd blast that sucker with a GPU radix sort 🔫 haha

With unique keys, some sort of trie would be great at storing a sparse index mapping, provided there is some clustering.

**Reply by elalish** [reply]
Yeah, plus the sorting is just so we can binary search it - really it's effectively a hash table. As I recall one time we tried to replace it with an actual hash table, but for some reason it performed worse, so we didn't bother. 

I think the real solution, which should decrease memory usage a lot and probably speed things up as well is refactor it so we don't need `SpareIndices` at all - just compute the shadow functions internally instead of storing them, even if it is a bit of duplicate compute. See some discussion of this idea [here](https://github.com/elalish/manifold/pull/931#issuecomment-2353465022).

**Reply by wtholliday** [reply]
What are shadow functions?

I looked inside `SparseIndices` returned by `Collider::Collisions` and the `p`s are not unique. So I don't see how one can make the memory usage for `SparseIndices` asymptotically better, and what you're saying about getting rid of `SparseIndices` altogether makes sense, even if I don't really understand the details.

**Reply by pca006132** [reply]
Those shadow functions are collision checks, but in 2D projections. The problem we got here is that our collider is not giving precise collision pairs, instead what we got is just bounding box intersections. Many of those collision pairs are not real collisions and can be removed.

I'm thinking if we need all 3 halfedges in p1q1 though, because we know that if the first halfedge is included, the other two will follow. If we can change it to include only 1 halfedge, we may reduce the memory usage of boolean3 a bit without too much changes?

**Reply by wtholliday** [reply]
> Those shadow functions are collision checks, but in 2D projections.

Do you mean it's projecting bounding boxes into some plane and then returning all the collisions?

**Reply by elalish** [reply]
Yeah, but I think I'm probably the only one in a position to attack that particular refactor. I'd like to get to it once we get v3.0 out. If you'd like to have a look at GLM in the meanwhile, that would probably be a great place to help.

**Reply by wtholliday** [reply]
Yeah I wasn't going to attempt that refactor. Was just hoping I could figure out some decent improvement to `SparseIndices`.

What if you had an array indexed by the `p`s of head indices of singly linked lists? Put a bunch of the q's in each list node. Similar to a hash table.

I'll take a look at the GLM stuff :)

**Comment by wtholliday** [comment]
FWIW `SparseIndices::Copy` doesn't seem like something that would benefit from parallelization within the `for_each` (too little calculation per job and it's just going to be memory bandwidth limited)

https://github.com/elalish/manifold/blob/31d6da097335786ab1a163866835412abc76f34e/src/utilities/include/manifold/sparse.h#L72

**Reply by pca006132** [reply]
Yes, it probably doesn't benefit much from parallelization here. In general, I think copy operations will not benefit much from parallelization... but probably need more profiling for this on different machines.

**Reply by pca006132** [reply]
Btw, you can try to run https://github.com/elalish/manifold/blob/master/extras/stl_test.cpp to see what algorithms are not having improvement with parallelization on your target device. Even copy can have some improvement when run in multiple threads, just that it may not be significant to warrant the additional CPU usage...
